File: golang/tests/pkg/integration/mobile_beta_screenshot_integration_test.go
package integration

import (
	"bytes"
	"os"
	"testing"
	"time"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestMobileBetaScreenshotPNG(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Skip("Skipping integration test: AGENTBAY_API_KEY not set")
	}

	agentBay, err := agentbay.NewAgentBay(apiKey)
	require.NoError(t, err, "Failed to create AgentBay client")

	sessionParams := &agentbay.CreateSessionParams{
		ImageId: "imgc-0ab5ta4mn31wth5lh",
	}
	sessionResult, err := agentBay.Create(sessionParams)
	require.NoError(t, err, "Failed to create session")
	require.NotNil(t, sessionResult.Session, "Session should be created")

	session := sessionResult.Session
	defer func() {
		_, _ = session.Delete()
	}()

	time.Sleep(15 * time.Second)

	cmds := []string{
		"wm size 720x1280",
		"wm density 160",
	}
	for _, c := range cmds {
		r, err := session.Command.ExecuteCommand(c, 10000)
		require.NoError(t, err)
		require.True(t, r.Success, "Command failed: %s error=%s", c, r.ErrorMessage)
	}

	start := session.Mobile.StartApp("monkey -p com.android.settings 1", "", "")
	require.NotEmpty(t, start.RequestID)
	require.Empty(t, start.ErrorMessage)
	time.Sleep(2 * time.Second)

	s := session.Mobile.BetaTakeScreenshot()
	assert.NotEmpty(t, s.RequestID)
	assert.True(t, s.Success, "beta screenshot failed: %s", s.ErrorMessage)
	assert.Equal(t, "png", s.Format)
	assert.NotNil(t, s.Width)
	assert.NotNil(t, s.Height)
	assert.Greater(t, *s.Width, 0)
	assert.Greater(t, *s.Height, 0)
	assert.True(t, bytes.HasPrefix(s.Data, []byte{0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a}))
	assert.Greater(t, len(s.Data), 8)

	ls := session.Mobile.BetaTakeLongScreenshot(2, "png")
	assert.NotEmpty(t, ls.RequestID)
	assert.True(t, ls.Success, "beta long screenshot failed: %s", ls.ErrorMessage)
	assert.Equal(t, "png", ls.Format)
	assert.NotNil(t, ls.Width)
	assert.NotNil(t, ls.Height)
	assert.Greater(t, *ls.Width, 0)
	assert.Greater(t, *ls.Height, 0)
	assert.True(t, bytes.HasPrefix(ls.Data, []byte{0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a}))
	assert.Greater(t, len(ls.Data), 8)
}

func TestMobileBetaLongScreenshotJPEGQuality(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Skip("Skipping integration test: AGENTBAY_API_KEY not set")
	}

	agentBay, err := agentbay.NewAgentBay(apiKey)
	require.NoError(t, err, "Failed to create AgentBay client")

	sessionParams := &agentbay.CreateSessionParams{
		ImageId: "imgc-0ab5ta4mn31wth5lh",
	}
	sessionResult, err := agentBay.Create(sessionParams)
	require.NoError(t, err, "Failed to create session")
	require.NotNil(t, sessionResult.Session, "Session should be created")

	session := sessionResult.Session
	defer func() {
		_, _ = session.Delete()
	}()

	time.Sleep(15 * time.Second)

	cmds := []string{
		"wm size 720x1280",
		"wm density 160",
	}
	for _, c := range cmds {
		r, err := session.Command.ExecuteCommand(c, 10000)
		require.NoError(t, err)
		require.True(t, r.Success, "Command failed: %s error=%s", c, r.ErrorMessage)
	}

	start := session.Mobile.StartApp("monkey -p com.android.settings 1", "", "")
	require.NotEmpty(t, start.RequestID)
	require.Empty(t, start.ErrorMessage)
	time.Sleep(2 * time.Second)

	invalid := session.Mobile.BetaTakeLongScreenshot(2, "jpeg", 0)
	assert.False(t, invalid.Success)
	assert.Contains(t, invalid.ErrorMessage, "invalid quality")
	assert.Equal(t, "jpeg", invalid.Format)

	high := session.Mobile.BetaTakeLongScreenshot(2, "jpeg", 95)
	low := session.Mobile.BetaTakeLongScreenshot(2, "jpeg", 10)

	assert.NotEmpty(t, high.RequestID)
	assert.True(t, high.Success, "beta long screenshot (jpeg, q=95) failed: %s", high.ErrorMessage)
	assert.Equal(t, "jpeg", high.Format)
	assert.NotNil(t, high.Width)
	assert.NotNil(t, high.Height)
	assert.Greater(t, *high.Width, 0)
	assert.Greater(t, *high.Height, 0)
	assert.True(t, bytes.HasPrefix(high.Data, []byte{0xff, 0xd8, 0xff}))
	assert.Greater(t, len(high.Data), 3)

	assert.NotEmpty(t, low.RequestID)
	assert.True(t, low.Success, "beta long screenshot (jpeg, q=10) failed: %s", low.ErrorMessage)
	assert.Equal(t, "jpeg", low.Format)
	assert.NotNil(t, low.Width)
	assert.NotNil(t, low.Height)
	assert.Greater(t, *low.Width, 0)
	assert.Greater(t, *low.Height, 0)
	assert.True(t, bytes.HasPrefix(low.Data, []byte{0xff, 0xd8, 0xff}))
	assert.Greater(t, len(low.Data), 3)

	assert.Greater(t, len(high.Data), len(low.Data))
}

File: typescript/tests/integration/mobile-beta-screenshot.integration.test.ts
// ci-stable
/**
 * Integration tests for Mobile beta screenshot APIs.
 *
 * Note: This test creates a real session. Do not run concurrently.
 */

import { AgentBay } from "../../src/agent-bay";
import { log } from "../../src/utils/logger";

describe("Mobile beta screenshot integration tests", () => {
  const apiKey = process.env.AGENTBAY_API_KEY;
  let agentBay: AgentBay;

  beforeEach(() => {
    if (apiKey) {
      agentBay = new AgentBay({ apiKey });
    }
  });

  test("should capture screenshot as PNG bytes", async () => {
    if (!apiKey) {
      log("Skipping test: AGENTBAY_API_KEY not set");
      return;
    }

    const sessionResult = await agentBay.create({
      imageId: "mobile-use-android-12-gw",
    });
    if (!sessionResult.success || !sessionResult.session) {
      throw new Error(
        `Failed to create session: ${sessionResult.errorMessage || ""}`
      );
    }
    const session = sessionResult.session!;

    try {
      await new Promise((resolve) => setTimeout(resolve, 15000));

      const cmds = ["wm size 720x1280", "wm density 160"];
      for (const c of cmds) {
        const r = await session.command.executeCommand(c, 10000);
        if (!r.success) {
          throw new Error(
            `command failed: ${c}, error=${r.errorMessage}, output=${r.output}`
          );
        }
      }

      const start = await session.mobile.startApp(
        "monkey -p com.android.settings 1"
      );
      expect(start.success).toBe(true);
      await new Promise((resolve) => setTimeout(resolve, 2000));

      const s = await session.mobile.betaTakeScreenshot();
      if (!s.success) {
        throw new Error(
          `betaTakeScreenshot failed: requestId=${s.requestId || ""}, error=${
            s.errorMessage || ""
          }`
        );
      }
      expect(s.requestId).toBeDefined();
      expect(s.type).toBe("image");
      expect(s.mimeType).toBe("image/png");
      expect(s.data.length).toBeGreaterThan(8);
      expect(
        Buffer.from(s.data)
          .slice(0, 8)
          .equals(Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]))
      ).toBe(true);
      expect(typeof s.width).toBe("number");
      expect(typeof s.height).toBe("number");
      expect((s.width as number) > 0).toBe(true);
      expect((s.height as number) > 0).toBe(true);
    } finally {
      await session.delete();
    }
  }, 60000);

  test("should capture screenshot as JPEG bytes", async () => {
    if (!apiKey) {
      log("Skipping test: AGENTBAY_API_KEY not set");
      return;
    }

    const sessionResult = await agentBay.create({
      imageId: "mobile-use-android-12-gw",
    });
    if (!sessionResult.success || !sessionResult.session) {
      throw new Error(
        `Failed to create session: ${sessionResult.errorMessage || ""}`
      );
    }
    const session = sessionResult.session!;

    try {
      await new Promise((resolve) => setTimeout(resolve, 15000));

      const cmds = ["wm size 720x1280", "wm density 160"];
      for (const c of cmds) {
        const r = await session.command.executeCommand(c, 10000);
        if (!r.success) {
          throw new Error(
            `command failed: ${c}, error=${r.errorMessage}, output=${r.output}`
          );
        }
      }

      const start = await session.mobile.startApp(
        "monkey -p com.android.settings 1"
      );
      expect(start.success).toBe(true);
      await new Promise((resolve) => setTimeout(resolve, 2000));

      const s = await (session.mobile as any).betaTakeScreenshot("jpeg");
      if (!s.success) {
        throw new Error(
          `betaTakeScreenshot failed: requestId=${s.requestId || ""}, error=${
            s.errorMessage || ""
          }`
        );
      }
      expect(s.requestId).toBeDefined();
      expect(s.type).toBe("image");
      expect(s.mimeType).toBe("image/jpeg");
      expect(s.data.length).toBeGreaterThan(3);
      expect(
        Buffer.from(s.data)
          .slice(0, 3)
          .equals(Buffer.from([0xff, 0xd8, 0xff]))
      ).toBe(true);
      expect(typeof s.width).toBe("number");
      expect(typeof s.height).toBe("number");
      expect((s.width as number) > 0).toBe(true);
      expect((s.height as number) > 0).toBe(true);
    } finally {
      await session.delete();
    }
  }, 60000);

  test("should capture long screenshot as PNG bytes", async () => {
    if (!apiKey) {
      log("Skipping test: AGENTBAY_API_KEY not set");
      return;
    }

    const sessionResult = await agentBay.create({
      imageId: "mobile-use-android-12-gw",
    });
    if (!sessionResult.success || !sessionResult.session) {
      throw new Error(
        `Failed to create session: ${sessionResult.errorMessage || ""}`
      );
    }
    const session = sessionResult.session!;

    try {
      await new Promise((resolve) => setTimeout(resolve, 15000));

      const cmds = ["wm size 720x1280", "wm density 160"];
      for (const c of cmds) {
        const r = await session.command.executeCommand(c, 10000);
        if (!r.success) {
          throw new Error(
            `command failed: ${c}, error=${r.errorMessage}, output=${r.output}`
          );
        }
      }

      const start = await session.mobile.startApp(
        "monkey -p com.android.settings 1"
      );
      expect(start.success).toBe(true);
      await new Promise((resolve) => setTimeout(resolve, 2000));

      const nav = await session.command.executeCommand(
        "am start -a android.settings.SETTINGS",
        10000
      );
      if (!nav.success) {
        throw new Error(
          `command failed: am start -a android.settings.SETTINGS, error=${nav.errorMessage}, output=${nav.output}`
        );
      }
      await new Promise((resolve) => setTimeout(resolve, 2000));

      const ls = await session.mobile.betaTakeLongScreenshot(2, "png");
      if (!ls.success) {
        throw new Error(
          `betaTakeLongScreenshot failed: requestId=${
            ls.requestId || ""
          }, error=${ls.errorMessage || ""}`
        );
      }
      expect(ls.requestId).toBeDefined();
      expect(ls.type).toBe("image");
      expect(ls.mimeType).toBe("image/png");
      expect(ls.data.length).toBeGreaterThan(8);
      expect(
        Buffer.from(ls.data)
          .slice(0, 8)
          .equals(Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]))
      ).toBe(true);
      expect(typeof ls.width).toBe("number");
      expect(typeof ls.height).toBe("number");
      expect((ls.width as number) > 0).toBe(true);
      expect((ls.height as number) > 0).toBe(true);
    } finally {
      await session.delete();
    }
  }, 60000);

  test("should capture long screenshot as JPEG bytes with quality", async () => {
    if (!apiKey) {
      log("Skipping test: AGENTBAY_API_KEY not set");
      return;
    }

    const sessionResult = await agentBay.create({
      imageId: "mobile-use-android-12-gw",
    });
    if (!sessionResult.success || !sessionResult.session) {
      throw new Error(
        `Failed to create session: ${sessionResult.errorMessage || ""}`
      );
    }
    const session = sessionResult.session!;

    try {
      await new Promise((resolve) => setTimeout(resolve, 15000));

      const cmds = ["wm size 720x1280", "wm density 160"];
      for (const c of cmds) {
        const r = await session.command.executeCommand(c, 10000);
        if (!r.success) {
          throw new Error(
            `command failed: ${c}, error=${r.errorMessage}, output=${r.output}`
          );
        }
      }

      const start = await session.mobile.startApp(
        "monkey -p com.android.settings 1"
      );
      expect(start.success).toBe(true);
      await new Promise((resolve) => setTimeout(resolve, 2000));

      const nav = await session.command.executeCommand(
        "am start -a android.settings.SETTINGS",
        10000
      );
      if (!nav.success) {
        throw new Error(
          `command failed: am start -a android.settings.SETTINGS, error=${nav.errorMessage}, output=${nav.output}`
        );
      }
      await new Promise((resolve) => setTimeout(resolve, 2000));

      const invalid = await session.mobile.betaTakeLongScreenshot(2, "jpeg", 0);
      expect(invalid.success).toBe(false);
      expect(invalid.errorMessage).toContain("Invalid quality");

      const high = await session.mobile.betaTakeLongScreenshot(2, "jpeg", 95);
      const low = await session.mobile.betaTakeLongScreenshot(2, "jpeg", 10);

      if (!high.success) {
        throw new Error(
          `betaTakeLongScreenshot (jpeg, q=95) failed: requestId=${
            high.requestId || ""
          }, error=${high.errorMessage || ""}`
        );
      }
      if (!low.success) {
        throw new Error(
          `betaTakeLongScreenshot (jpeg, q=10) failed: requestId=${
            low.requestId || ""
          }, error=${low.errorMessage || ""}`
        );
      }

      expect(high.requestId).toBeDefined();
      expect(high.type).toBe("image");
      expect(high.mimeType).toBe("image/jpeg");
      expect(high.data.length).toBeGreaterThan(3);
      expect(
        Buffer.from(high.data)
          .slice(0, 3)
          .equals(Buffer.from([0xff, 0xd8, 0xff]))
      ).toBe(true);
      expect(typeof high.width).toBe("number");
      expect(typeof high.height).toBe("number");
      expect((high.width as number) > 0).toBe(true);
      expect((high.height as number) > 0).toBe(true);

      expect(low.requestId).toBeDefined();
      expect(low.type).toBe("image");
      expect(low.mimeType).toBe("image/jpeg");
      expect(low.data.length).toBeGreaterThan(3);
      expect(
        Buffer.from(low.data)
          .slice(0, 3)
          .equals(Buffer.from([0xff, 0xd8, 0xff]))
      ).toBe(true);
      expect(typeof low.width).toBe("number");
      expect(typeof low.height).toBe("number");
      expect((low.width as number) > 0).toBe(true);
      expect((low.height as number) > 0).toBe(true);

      expect(high.data.length).toBeGreaterThan(low.data.length);
    } finally {
      await session.delete();
    }
  }, 90000);
});
